cmd/validators/main_test.go

```diff
@@ -0,0 +1,124 @@
+package main
+
+import (
+	"bytes"
+	"flag"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+)
+
+func TestMultisig(t *testing.T) {
+	// Create, sign, view and submit a multisignature request flow. This is
+	// not a unit test neither an integration test. It is nevertheless
+	// helpful.
+	tm := newTendermintServer(t)
+	defer tm.Close()
+
+	var out bytes.Buffer
+	args := []string{
+		"-power", "7",
+		"-pubkey", "j4JRVstX",
+		"-multisig", "5AE2C58796B0AD48FFE7602EAC3353488C859A2B",
+	}
+	if err := cmdMultisigNew(nil, &out, args); err != nil {
+		t.Fatalf("cannot create a multisig request: %s", err)
+	}
+	unsignedReq := out.Bytes()
+	out.Reset()
+	if err := cmdMultisigView(bytes.NewReader(unsignedReq), &out, nil); err != nil {
+		t.Fatalf("cannot view a multisig request: %s", err)
+	}
+	t.Logf("unsigned request: %s", out.String())
+
+	out.Reset()
+	args = []string{
+		"-tm", tm.URL,
+		"-key", "d34c1970ae90acf3405f2d99dcaca16d0c7db379f4beafcfdf667b9d69ce350d27f5fb440509dfa79ec883a0510bc9a9614c3d44188881f0c5e402898b4bf3c9",
+	}
+	if err := cmdMultisigSign(bytes.NewReader(unsignedReq), &out, args); err != nil {
+		t.Fatalf("cannot sign a multisig request: %s", err)
+	}
+
+	signedReq := out.Bytes()
+	out.Reset()
+	if err := cmdMultisigView(bytes.NewReader(signedReq), &out, nil); err != nil {
+		t.Fatalf("cannot view a multisig request: %s", err)
+	}
+	t.Logf("signed request: %s", out.String())
+
+	out.Reset()
+	args = []string{
+		"-tm", tm.URL,
+	}
+	if err := cmdMultisigSubmit(bytes.NewReader(signedReq), &out, args); err != nil {
+		t.Fatalf("cannot submit a multisig request: %s", err)
+	}
+}
+
+func newTendermintServer(t *testing.T) *httptest.Server {
+	t.Helper()
+
+	return httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if *logRequestFl {
+			b, _ := ioutil.ReadAll(r.Body)
+			r.Body = ioutil.NopCloser(bytes.NewReader(b))
+			t.Logf("tendermint request: %s %s: %s", r.Method, r.URL.Path, string(b))
+		}
+		switch {
+		case r.Method == "GET" && r.URL.Path == "/genesis":
+			io.WriteString(w, `
+				{
+					"jsonrpc": "2.0",
+					"id": "",
+					"result": {
+						"genesis": {
+							"chain_id": "test-chain-ZIYjN0",
+							"validators": [],
+							"app_state": {
+								"gconf": {
+									"cash:minimal_fee": {},
+									"cash:collector_address": "0000000000000000000000000000000000000000"
+								},
+								"update_validators": {
+									"addresses": [
+										"b1ca7e78f74423ae01da3b51e676934d9105f282",
+										"E28AE9A6EB94FC88B73EB7CBD6B87BF93EB9BEF0"
+									]
+								},
+								"multisig": [
+								{
+									"sigs": [
+										"b1ca7e78f74423ae01da3b51e676934d9105f282",
+										"E28AE9A6EB94FC88B73EB7CBD6B87BF93EB9BEF0"
+									],
+									"admin_threshold": 2,
+									"activation_threshold": 1
+								}
+								]
+							}
+						}
+					}
+				}
+			`)
+		case r.Method == "POST" && r.URL.Path == "/":
```

Comment:
I think there are many paths there... like `/broadcast_tx_commit`.
Also, don't you read the nonce from the chain?

Reply:
Maybe there is an easier way than mocking it all?
If not, then it is quite a bit of work to build a proper mock that will return nice results, and I would mock out on the bnsClient level, which has higher-level API

Reply:
Test that is provided is using only `POST /`.

I was aiming for the minimal server implementation that will allow passing the tests. I do not understand the tendermint protocol enough to try to provide a solid implementation.

I think the nonce is read from the chain with a fallback if it does not exist. I think this is also done with `POST /` because I did not log another request.

Below is the full test run output. Lines with `tendermint request:` are important:
```
$ go test github.com/iov-one/weave/cmd/validators/... -logrequest -v
2019/02/11 13:32:42 proto: duplicate proto type registered: common.KVPair
2019/02/11 13:32:42 proto: duplicate proto type registered: common.KI64Pair
=== RUN   TestMultisig
--- PASS: TestMultisig (0.00s)
    main_test.go:34: unsigned request: {
                "Sum": {
                        "SetValidatorsMsg": {
                                "validator_updates": [
                                        {
                                                "pubkey": {
                                                        "type": "ed25519",
                                                        "data": "j4JRVstX"
                                                },
                                                "power": 7
                                        }
                                ]
                        }
                }
        }
    main_test.go:68: tendermint request: GET /genesis: 
    main_test.go:68: tendermint request: POST /: {"jsonrpc":"2.0","id":"jsonrpc-client","method":"abci_query","params":{"data":"E28AE9A6EB94FC88B73EB7CBD6B87BF93EB9BEF0","height":"0","path":"/auth","prove":false}}
    main_test.go:50: signed request: {
                "signatures": [
                        {
                                "pubkey": {
                                        "Pub": {
                                                "Ed25519": "J/X7RAUJ36eeyIOgUQvJqWFMPUQYiIHwxeQCiYtL88k="
                                        }
                                },
                                "signature": {
                                        "Sig": {
                                                "Ed25519": "6EJlCaTdYE/dCptm7Jq/6drRS/e95iKTBaNfPcgf9k53VFiqsuXBPwGYPC7Htn1rK+tyYnVw/xr9W4Wy1tjvCA=="
                                        }
                                }
                        }
                ],
                "Sum": {
                        "SetValidatorsMsg": {
                                "validator_updates": [
                                        {
                                                "pubkey": {
                                                        "type": "ed25519",
                                                        "data": "j4JRVstX"
                                                },
                                                "power": 7
                                        }
                                ]
                        }
                }
        }
    main_test.go:68: tendermint request: POST /: {"jsonrpc":"2.0","id":"jsonrpc-client","method":"broadcast_tx_commit","params":{"tx":"EmgSIgogJ/X7RAUJ36eeyIOgUQvJqWFMPUQYiIHwxeQCiYtL88kiQgpA6EJlCaTdYE/dCptm7Jq/6drRS/e95iKTBaNfPcgf9k53VFiqsuXBPwGYPC7Htn1rK+tyYnVw/xr9W4Wy1tjvCNIDFwoVChEKB2VkMjU1MTkSBo+CUVbLVxAH"}}
PASS
ok      github.com/iov-one/weave/cmd/validators 0.021s
```

Reply:
Good call, yes
> POST /: {"jsonrpc":"2.0","id":"jsonrpc-client","method":"abci_query","params":{"data":"E28AE9A6EB94FC88B73EB7CBD6B87BF93EB9BEF0","height":"0","path":"/auth","prove":false}}
is the line I was thinking of, didn't realize which protocol uses which path (there is URI/HTTP GET, JsonRPC/HTTP POST, and Websockets).

Reply:
I will do one more pass and then give you a :+1: